projectile-rails.el

```diff
@@ -970,28 +970,36 @@ The bound variable is \"filename\"."
 
 (defun projectile-rails-find-constant (name)
   (let* ((code-dirs (-filter #'f-exists? (-map #'projectile-rails-expand-root (projectile-rails--code-directories))))
-         (file-name (format "%s.rb" (projectile-rails-declassify name)))
          (list-parent-dirs (lambda (some-file)
                              (let ((parent-dirs '()))
                                (f-traverse-upwards (lambda (parent-dir)
                                                      (push (f-canonical parent-dir) parent-dirs)
                                                      (equal (f-slash (f-canonical (projectile-rails-root))) (f-slash (f-canonical parent-dir))))
                                                    (f-dirname some-file))
                                parent-dirs)))
-         (lookup-dirs (-flatten (list
-                                 ;; Look in current file namespace
-                                 (f-no-ext buffer-file-name)
-                                 ;; Look in local namespace hierarchy
-                                 (funcall list-parent-dirs buffer-file-name)
-                                 ;; Look in code directories
-                                 code-dirs)))
-         (lookup-paths (--map (f-join it file-name)
+         (file-name (format "%s.rb" (projectile-rails-declassify name)))
+         (lookup-dirs (if (f-absolute? file-name)
+                          ;; If top-level constant (e.g. ::Classname), i.e. derived filename (/classname) starts with a "/", then:
+                          ;; Look only in code directories
+                          code-dirs
+                        (-flatten (list
+                                   ;; Otherwise (relative constant):
+                                   ;; 1. Look in current file namespace
+                                   (f-no-ext buffer-file-name)
+                                   ;; 2. Look in local namespace hierarchy
+                                   (funcall list-parent-dirs buffer-file-name)
+                                   ;; 3. Look in code directories
+                                   code-dirs))))
+         ;; Strip leading "/" if present before generating lookup paths because it messes with f-join)
+         (relative-file-name (if (f-absolute? file-name)
+                                 (substring file-name 1)
+                               file-name))
+         (lookup-paths (--map (f-join it relative-file-name)
```

Comment:
While I'm not a big fan of ifs myself (well I don't hate them but I also like to stop and think when I see one, see if I can replace it with something more maintainable, e.g. with polymorphism in OO), I reckon this one is much easier to read and reason about than that regex.

Besides, the same `if` appears right above and repeating it makes it easier to see that they're related – both handling the same "absolute namespace" scenario – which using a regex for one would obfuscate.

What I would really like is to do is DRY-up those 2 ifs into one, but the only way I can think of would add at least 2 levels of nesting, making the neat `let*` sequence messier and possibly less approachable.

Anyway, let me know if you feel strongly about replacing the if with a regex, I'll be happy to oblige.